sql/item_strfunc.cc

```diff
@@ -1302,6 +1302,133 @@ bool Item_func_replace::fix_length_and_dec()
 }
 
 
+bool Item_func_sformat::fix_length_and_dec()
+{
+  ulonglong char_length= 0;
+  
+  if (agg_arg_charsets_for_string_result(collation, args, arg_count))
+    return TRUE;
+  
+  for (uint i=0 ; i < arg_count ; i++)
+    char_length+= args[i]->max_char_length() * 10; // temporal solution
+  
+  fix_char_length_ulonglong(char_length);
+  return FALSE;
+}
+
+longlong Item_func_sformat::locate(String *a, String *b, longlong offset)
+{
+  /* must be longlong to avoid truncation */
+  longlong   start=  0;
+  my_match_t match;
+  
+  if (offset > 0)
+  {
+    start= offset;
+    
+    if ((start <= 0) || (start > a->length()))
+      return 0;
+    offset--;
+    start--;
+    
+    /* start is now sufficiently valid to pass to charpos function */
+    start= a->charpos((int) start);
+    
+    if (start + b->length() > a->length())
+      return 0;
+  }
+  
+  if (!b->length())       // Found empty string at start
+    return start + 1;
+  
+  if (!cmp_collation.collation->instr(a->ptr() + start,
+                                      (uint) (a->length() - start),
+                                      b->ptr(), b->length(),
+                                      &match, 1))
+    return 0;
+  return (longlong) match.mb_len + offset + 1;
+}
+
+String *Item_func_sformat::substr(String *res, longlong start, longlong length)
+{
+  longlong tmp_length;
+
+  if (length <= 0)
+    return make_empty_result();
+  
+  /* Assumes that the maximum length of a String is < INT_MAX32. */
+  /* Set here so that rest of code sees out-of-bound value as such. */
+  if (length > INT_MAX32)
+    length= INT_MAX32;
+  
+  /* Assumes that the maximum length of a String is < INT_MAX32. */
+  if (start > INT_MAX32)
+    return make_empty_result();
+  
+  start= ((start < 0) ? res->numchars() + start : start - 1);
+  start= res->charpos((int) start);
+  if ((start < 0) || ((uint) start + 1 > res->length()))
+    return make_empty_result();
+  
+  length= res->charpos((int) length, (uint32) start);
+  tmp_length= res->length() - start;
+  length= MY_MIN(length, tmp_length);
+  
+  if (!start && (longlong) res->length() == length)
+    return res;
+  tmp_value.set(*res, (uint32) start, (uint32) length);
+  return &tmp_value;
+}
+
+String *Item_func_sformat::val_str(String *res)
+{
+  DBUG_ASSERT(fixed());
+  uint     i=1;
+  String  *format= NULL;
+  String  *tmp= NULL;
+  String   keys;
+  longlong index= 0;
+  longlong prev= 1;
+  
+  res->length(0);
+  res->set_charset(collation.collation);
+  
+  null_value= false;
+  if (!(format= args[0]->val_str(&tmp_value)))
+    goto null;
+  
+  keys.set("{}", 2, format->charset());
+  index= locate(format, &keys, index);
+  while (index != 0) {
+    // Add a part of the format
+    tmp= substr(format, prev, index-prev);
+    res->append(*tmp);
+    // Add the parameter
+    if (i < arg_count) {
```

Comment:
I won't comment on indentation now. Try to follow the existing coding style, but don't strain yourself too much over it. This can be fixed later, before pushing. Better not to slow down now because of this.